### DISCUSSION TITLE
Remove Executor Coupling to Core Airlfow Code Base

### DISCUSSION THREAD

**Discussion by o-nikolas** [apache/airflow#27241] [Ideas]
Hello Airflow folks!

# Summary

I have recently been investigating areas where the Executor interface/API is leaking into core Airflow code. Executors, as they are now, are quite pluggable/extensible but there remains several places in the core Airflow logic (e.g. scheduler, backfill, settings/config, etc) where the code is opinionated about _specific_ Executors. However, the core Airflow logic should not need to know the properties of any one specific Executor, it should be able to interact with a generic interface which all Executors implement. Put another way, if one is looking to add a new Executor, they should **not** have to modify any core Airflow code/logic.

A concrete example helps illustrate the point:
https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/configuration.py#L412-L419

Here you can see code in Configuration.py maintaining a set of specific Executors that do not support SQLite. So if someone was to add a new Executor, which does not support SQLite, then they would need to modify this piece of core Airflow code (and the other places where SQLite compatibility are checked). This is Executor coupling!

# Goal
With this Discussion thread I hope to:

1. List the points of Executor coupling that I've found whilst digging around in the Airflow codebase and request others to add any they've seen as well.
1. Get the temperature of the community on their level of enthusiasm and motivation to remove (or not) each of the various points of Executor coupling.
1. Discuss options to tackle each point of coupling (some are easier than others). As I'm sure folks will have many differing opinions on how to do this, and that's a good thing :smiley:

# Areas of Executor Coupling and Proposals to Remediate
## 1) Local Executor Compatibility Checks
Executors that run locally have certain limitations and requirements, many of which are currently hardcoded in core Airflow code. To add a new Executor, that would run locally, one would be required to change this core Airflow code.
### Examples
- **1a**) Whether or not to start the serve_logs sub-process, currently hardcoded to LocalExecutor and SequentialExecutor: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/cli/commands/scheduler_command.py#L93
- **1b**) When running in standalone mode, asserts a local Executor is being used, currently hardcoded to LocalExecutor and SequentialExecutor: https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/cli/commands/standalone_command.py#L156-L158
- **1c**) Local Executors need a copy of Airflow configuration created in tmp, currently hardcoded to LocalExecutor and SequentialExecutor: https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/jobs/backfill_job.py#L505-L509
### Proposal
A static method or attribute on the Executor class which can be checked by core code. There is a precedent already set with the `supports_ad_hoc_ti_run` attribute:
https://github.com/apache/airflow/blob/fb741fd87254e235f99d7d67e558dafad601f253/airflow/executors/kubernetes_executor.py#L435 https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/www/views.py#L1735-L1737
Static methods and class attributes are convenient for this purpose since the Executor class would not need to be instantiated but are difficult if not impossible to enforce implementation as one would with instance methods.

##### Note: that while the usage of `supports_ad_hoc_ti_run` is a good example of uncoupled Executor logic, the logging msg in the above example does technically have some coupling because it mentions specific Executors. It should be updated to something like the below:
 ```python
 if not getattr(executor, "supports_ad_hoc_ti_run", False): 
     msg = f"{executor.__class__.__name__} does not support ad hoc TI runs" 
     return redirect_or_json(origin, msg, "error", 400) 
```

## 2) Single Threaded Executors (SQLite support)
Some Executors, currently a subset of the local Executors, run in a single threaded fashion and have certain limitations and requirements, many of which are hardcoded. To add a new single threaded Executor would require changes to core Airflow code.
Note: This coupling often shows up with SQLite compatibility checks since it does not support multiple connections.
### Examples
- **2a**) SQLite check done in configuration.py: https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/configuration.py#L412-L419
- **2b**) When running in standalone mode SQLite compatibility is checked: https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/cli/commands/standalone_command.py#L160-L165
- **2c**) Sensors in `poke` mode can block execution of DAGs when running with single process Executors, currently hardcoded to DebugExecutor (although should also include SequentialExecutor): https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/sensors/base.py#L243
### Proposal
Similar to the proposal for coupling 1, a static method or attribute for single threaded compatibility can be added to the BaseExecutor interface, then implemented by child classes, then checked by core Airflow code instead of hardcoding specific Executors.

## 3) Misc Compatibility Checks
Coupling around other compatibility checks, similar to coupling 1 and 2, but which don't fall into either category.
### Examples
- **3a**) Some Executors support pickling, some don’t, there is logic in both backfill and scheduler jobs that checks before submitting tasks to the Executor queue. Applies to DaskExecutor, LocalExecutor and SequentialExecutor as of now: https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/jobs/backfill_job.py#L825-L829 https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/jobs/scheduler_job.py#L703-L704
- **3b**) Sentry has a celery integration, if the Executor is Celery, the Sentry integration is imported and configured: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/sentry.py#L87
  - Sentry at the moment only supports Celery, so this particular coupling is low priority, but should still be fixed.
### Proposal
The above could be resolved in a similar way to couplings 1 and 2.

## 4) Executor Coupling in Logging
The FileTaskHandler logging class has a special branch for Kubernetes, to fetch logs from the pod (Note: this is currently hardcoded to KubernetesExecutor so this would fail to run for the CeleryKubernetesExecutor or LocalKubernetesExecutor).
### Examples
- **4a**) See code: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/utils/log/file_task_handler.py#L157-L192
### Proposal
Adding a `get_task_logs(ti: TaskInstance | None)` method on the `BaseExecutor` class, which Executor subclasses could implement is one viable approach, since other Executors could certainly make use of such a paradigm, particularly containerized or cloud based executors.

## 5) Executor Specific CLI Commands
Some Executors have their own first class CLI commands (now that’s hardcoding/coupling!) which setup or modify various components related to that Executor.
### Examples
- **5a**) Celery Executor commands: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/cli/cli_parser.py#L1689-L1734
- **5b**) Kubernetes Executor commands: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/cli/cli_parser.py#L1754-L1771
- **5c**) Default CLI parser has hardcoded logic for Celery and Kubernetes Executors specifically: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/cli/cli_parser.py#L63-L99
### Proposal
This is a behaviour that Airflow has had for quite some time, but is a form of coupling that feels odd when looking in retrospect. However, that is not to say they are not useful for our users. So while simply deprecating them (making Airflow CLI unopinionated about Executor management) is possible, it does not seem like the right solution (though I would like to hear some thoughts on this!). If we are committed to this paradigm, it could be a very clean and straightforward solution to update the BaseExecutor interface with a pluggable mechanism to vend CLI `GroupCommands` and parsers. Executor subclasses would then implement these methods, if applicable, which would then be called to fetch commands and parsers from within Airflow Core cli parser code. We would then migrate the existing Executor CLI code from cli_parser to the respective Executor class.

Pseudo-code example for vending `GroupCommand`s:
```python
# Existing code in cli_parser.py
...
airflow_commands: List[CLICommand] = [
GroupCommand(
        name='dags',
        help='Manage DAGs',
        subcommands=DAGS_COMMANDS,
    ),
    ...
]
# New code to add groups vended by executor classes
executor_cls, _ = ExecutorLoader.import_executor_cls(conf.get('core', 'EXECUTOR'))
airflow_commands.append(executor_cls.get_cli_group_commands())
...
```

## 6) Misc Executor Coupling in UI
There are a few pages in the Airflow UI which are specific to particular Executors.
### Examples
- **6a**) Kube yaml can be rendered in the UI for KubernetesExecutor and CeleryKubernetesExecutor, at the `/rendered-k8s` url.
Views code: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/www/views.py#L1480
Associated TI code: https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/models/taskinstance.py#L2161
- **6b**) A warning popup is show in the UI when using the SequentialExecutor.
Views code: https://github.com/apache/airflow/blob/26f94c5370587f73ebd935cecf208c6a36bdf9b6/airflow/www/views.py#L532
Associated message in dags.html: https://github.com/apache/airflow/blob/fb741fd87254e235f99d7d67e558dafad601f253/airflow/www/templates/airflow/dags.html#L115-L120
### Proposal
One could imagine a pluggable mechanism, similar to coupling 5, for Executor subclasses to vend UI views to then be used in views.py. This would allow all Executors to have UI components without needing to modify core Airflow code. There are not many instance of this coupling and it would be a more involved implementation which might not be used often. It isn't as high priority as other coupling types, but should still be addressed. There are also projects underway to migrate Airflow UI to a different framework, which would likely affect how this coupling would be solved.

## 7) Specific Executors as Defaults
There are places in core Airflow code where specific Executors are the default value for some specific workflow or action.
### Examples
- **7a**) The `DAG.run()` method, when `local=True` and no other Executor is provided, is hardcoded to use local Executor: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/models/dag.py#L2398
### Proposal
I'd argue that in this case, and other cases like it, using a specific Executor as a default does not count as strict Executor coupling. No part of the Executor interface is being leaked. This type of coupling doesn't fail the test of "if I wanted to add a new Executor, would I need to update this code?". In these cases I'm inclined to leave these defaults in place and simply add ways for users to override which Executor is used where applicable and not already done so (like example 7a).

# Conclusion
I'd like to hear any feedback on the above and open Issues for each coupling type so that myself or anyone else in the community who's interested can be assigned the task and begin work. Thanks for your time!

**Comment by potiuk** [comment]
I love the clarity and detail level. I think it is pretty close (if not ready ) to be put as AIP proposal

**Comment by eladkal** [comment]
Very clear indeed!
I have a question about (6) we have cases where UI logic is shared between more than 1 executor for example in Run button which relevant for Celery, Kubernetes, CeleryKubernetes executors. How would that work in this case?

**Reply by o-nikolas** [reply]
Thanks for the feedback!

I didn't find any direct coupling there in my search through the code base. Do you mind linking to the code in question so I can investigate further?

**Reply by eladkal** [reply]
https://github.com/apache/airflow/blob/9ab1a6a3e70b32a3cddddf0adede5d2f3f7e29ea/airflow/www/views.py#L1753-L1755
It's associated
![Screen Shot 2022-10-31 at 20 56 17](https://user-images.githubusercontent.com/45845474/199089858-3cefa03a-9451-46e3-a690-8895a23f36b0.png)
 with the Run button

**Reply by pierrejeambrun** [reply]
Checking for `supports_ad_hoc_ti_run` seem good to me. I believe the easiest solution would be to change the message. (No enumeration of executors supporting the feature)

**Reply by eladkal** [reply]
I'm confused about how the logic would wrap this button...
If we decouple the executor from the UI you need somehow to be able to set conditional over specific buttons (I guess this is something we can debate in a more advanced phase)

**Reply by o-nikolas** [reply]
Yupp, @pierrejeambrun is right here. The `supports_ad_hoc_ti_run` is actually a good example of uncoupled exexcutor logic (see the proposal for Coupling 1). Each executor can choose to opt in or out of that feature by changing it's class implementation and then this code you linked checks if the attribute exists.

The only thing that needs updating is the msg string itself. Which I actually had in one draft of my doc but removed it since it was too deep into implementation detail. But essentially what that should be is something like this (the msg string needs workshoping, but just for the gist):
```python
 if not getattr(executor, "supports_ad_hoc_ti_run", False): 
     msg = f"{executor.__class__.__name__} does not support ad hoc TI runs" 
     return redirect_or_json(origin, msg, "error", 400) 
```

**Reply by o-nikolas** [reply]
I'll update this Discussion, and be sure to cover it in the AIP.

**Comment by pierrejeambrun** [comment]
Totally makes sense to me, thanks for the details.

I would love to help implementing some of it when we get there :)

**Comment by jon-evergreen** [comment]
Can this AIP also consider a way of implementing composed operators like the `LocalKubernetesExecutor` without those being hardcoded in [airflow/executors/executor_loader.py](https://github.com/apache/airflow/blob/main/airflow/executors/executor_loader.py) too?

I currently use a executor which uses ECS for executing tasks (I don't want the overhead of running a k8s cluster), but being able to switch some of those executions to a local executor via a similar implementation to `LocalKubernetesExecutor` would be really nice.

**Reply by o-nikolas** [reply]
I have a separate AIP coming down the pipe to handle the use case of more easily mixing and matching executors. Stay tuned for that! :smiley:
